### ISSUE TITLE
Gradle fails to pull in transitive dependency with classifier

### ISSUE THREAD

**Issue by TheMrMilchmann** [gradle/gradle#21526]
Gradle is currently failing to resolve a transitive dependency with a classifier.

The transitive dependency declaration from the POM:
```xml
<dependency>
    <groupId>net.java.dev.jna</groupId>
    <artifactId>jna</artifactId>
    <version>5.12.1</version>
    <classifier>jpms</classifier>
    <scope>runtime</scope>
</dependency>
```

and Gradle's `.module` file:
```json
{
  "group": "net.java.dev.jna",
  "module": "jna",
  "version": {
    "requires": "5.12.1"
  },
  "thirdPartyCompatibility": {
    "artifactSelector": {
      "name": "jna",
      "type": "jar",
      "classifier": "jpms"
    }
  }
},
```

### Expected Behavior

Gradle should resolve transitive dependencies with classifiers just fine.

### Current Behavior

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> Could not resolve all files for configuration ':testRuntimeClasspath'.
   > Could not find jna-platform-5.12.1-jpms (net.java.dev.jna:jna-platform:5.12.1).
     Searched in the following locations:
         https://repo.maven.apache.org/maven2/net/java/dev/jna/jna-platform/5.12.1/jna-platform-5.12.1-jpms
   > Could not find jna-5.12.1-jpms (net.java.dev.jna:jna:5.12.1).
     Searched in the following locations:
         https://repo.maven.apache.org/maven2/net/java/dev/jna/jna/5.12.1/jna-5.12.1-jpms
```

Gradle fails with the error message included above. Instinctively, I would assume that Gradle somehow loses the file extension in this case.

### Steps to Reproduce 

1. Clone https://github.com/TheMrMilchmann/MCVE/tree/gradle/21526-transitive-deps-classifier
2. Run `gradlew check`

### Your Environment

Build scan URL: https://scans.gradle.com/s/yfdjegfcwc6wa

**Comment by jbartok** [comment]
Observation (hint for the bug):
```
     Searched in the following locations:
         https://repo.maven.apache.org/maven2/net/java/dev/jna/jna/5.12.1/jna-5.12.1-jpms
```
is missing the extension from the URL.

**Comment by rpmcfong** [comment]
I just encountered this problem as well with a module that is internal to my company.

**Comment by mateuszkwiecinski** [comment]
It seems like I faced the same issue:
In my case, I changed the dependency format from `group:module:version:classifier` to gradle dsl - [link](https://github.com/usefulness/lazythreetenbp/pull/75)
Which resulted in missing `extension` in published `.module` file
From:
``` groovy
configuration(group:module:version:classifier)
```
https://repo1.maven.org/maven2/com/github/usefulness/lazythreetenbp/0.17.0/lazythreetenbp-0.17.0.module
To:
```groovy
configuration(libs.dependency) { artifact { classifier = 'classifier' } }
```
https://repo1.maven.org/maven2/com/github/usefulness/lazythreetenbp/0.18.0/lazythreetenbp-0.18.0.module

which resulted in a following diff:
![image](https://user-images.githubusercontent.com/36954793/199006615-182b3283-88b0-41d8-b601-59f33f20d6f6.png)

which is the same behaviour as described in the initial post, and I confirm the workaround posted above also works 🤷

**Comment by blaghed** [comment]
Can confirm also the same issue when moving to Gradle Version Catalogue approach.

In my case, this is in the declaration of a transitive dependency in "project A", which then fails to be resolved in "project B", so the workaround above could not be applied (as the dependency is not explicitely declared).

Instead, being painfully explicit in "project A" directly is what allowed me to work around the issue.

### Example

In Project A, the transitive library is declared:
```groovy
api 'org:foo:1.2.3' // OK
api(libs.foo) { artifact { classifier = 'custom' } } // NOK
api(libs.foo) { artifact { classifier = 'custom'; type = 'jar'; extension = 'jar' } } // OK
```

**Comment by big-guy** [comment]
Need to confirm if this was fixed by https://github.com/gradle/gradle/pull/28354

**Comment by mateuszkwiecinski** [comment]
I confirm [my case](https://github.com/gradle/gradle/issues/21526#issuecomment-1297011751) seems to be addressed starting from latest stable Gradle 8.8 👍 (+ I confirmed it still didn't work in 8.7) Thank you 🙏

**Comment by TheMrMilchmann** [comment]
I can confirm that this issue is fixed in Gradle 8.8. Thanks! Just one final note for documentation purposes: It might be necessary to run with `--refresh-dependencies` for the fix to take effect.
